File: cpp-alg/leecode/Q234_palindrome_linked_list.cpp
#include <iostream>
#include <vector>

struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

bool arraySolution(ListNode *head) {
  std::vector<int> vals;
  while (head != nullptr) {
    vals.emplace_back(head->val);
    head = head->next;
  }
  for (int i = 0, j = (int)vals.size() - 1; i < j; ++i, --j) {
    if (vals[i] != vals[j])
      return false;
  }
  return true;
}

ListNode *reverseListNode(ListNode *head) {
  ListNode *prev = nullptr;
  ListNode *current = head;
  while (current) {
    ListNode *next = current->next;
    current->next = prev;
    prev = current;
    current = next;
  }
  return prev;
}

bool slowFastPointer(ListNode *head) {
  ListNode *slow = head;
  ListNode *fast = head;

  while (fast->next && fast->next->next) {
    slow = slow->next;
    fast = fast->next->next;
  }

  ListNode *secondHalf = reverseListNode(slow->next);

  while (secondHalf) {
    if (head->val != secondHalf->val)
      return false;
    head = head->next;
    secondHalf = secondHalf->next;
  }

  return true;
}

bool isPalindrome(ListNode *head) {
  // array solution
  // copy to an array, then compare the first elem and last elem
  // like [1, 2, 2, 1], take array[0] and array[array.size() - 1) to compare
  // return arraySolution(head);

  // slowFastPointer solution
  // find the middle and reverse one of the parts using slow and fast pointer
  // reverse slow->next is because slow pointing to the end of the first half,
  // slow->next issecond half.
  return slowFastPointer(head);
}

int main() {

  ListNode *nodeG = new ListNode(1);
  ListNode *nodeF = new ListNode(2, nodeG);
  ListNode *nodeE = new ListNode(3, nodeF);
  ListNode *nodeD = new ListNode(3, nodeE);
  ListNode *nodeC = new ListNode(2, nodeD);
  ListNode *nodeB = new ListNode(1, nodeC);
  bool result = isPalindrome(nodeB);
  std::cout << result << std::endl;
}

File: python-alg/leecode/q234_palindrome_linked_list.py
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def copy_node(node):
    if node is None:
        return None
    new_node = ListNode(node.val)
    new_node.next = copy_node(node.next)
    return new_node


def palindrome_linked_list(node):
    head = copy_node(node)
    prev = None
    curr = node
    while curr:
        temp = curr.next
        curr.next = prev
        prev = curr
        curr = temp

    while prev and head:
        if prev.val != head.val:
            return False
        prev = prev.next
        head = head.next
    return True


def array_solution(node):
    vals = []
    head = copy_node(node)
    while head:
        vals.append(head.val)
        head = head.next
    i = 0
    j = len(vals) - 1
    while i < j:
        if vals[i] != vals[j]:
            return False
        i += 1
        j -= 1
    return True


nodeB = ListNode(2)
nodeA = ListNode(1)
nodeA.next = nodeB
# print(palindrome_linked_list(nodeA))
print(array_solution(nodeA))
